be/src/storage/compaction_scheduler.cpp

```diff
@@ -0,0 +1,273 @@
+            LOG(INFO) << "skip tablet:" << tablet->tablet_id()
+                      << " for limit of cumulative compaction task per disk. disk path:" << data_dir->path()
+                      << ", running num:" << num;
+            return false;
+        }
+        last_failure_ms = tablet->last_cumu_compaction_failure_time();
+    } else {
+        if (config::max_base_compaction_task >= 0 && task_num >= config::max_base_compaction_task) {
+            LOG(INFO) << "skip tablet:" << tablet->tablet_id()
+                      << " for base compaction limit:" << config::max_base_compaction_task;
+            return false;
+        }
+        std::unique_lock lk(tablet->get_base_lock(), std::try_to_lock);
+        if (!lk.owns_lock()) {
+            LOG(INFO) << "skip tablet:" << tablet->tablet_id() << " for base lock";
+            return false;
+        }
+        uint16_t num = CompactionManager::instance()->running_base_tasks_num_for_dir(data_dir);
+        if (config::base_compaction_num_threads_per_disk >= 0 && num >= config::base_compaction_num_threads_per_disk) {
+            LOG(INFO) << "skip tablet:" << tablet->tablet_id()
+                      << " for limit of base compaction task per disk. disk path:" << data_dir->path()
+                      << ", running num:" << num;
+            return false;
+        }
+        last_failure_ms = tablet->last_base_compaction_failure_time();
+    }
+    int64_t now_ms = UnixMillis();
+    if (now_ms - last_failure_ms <= config::min_compaction_failure_interval_sec * 1000) {
+        LOG(INFO) << "Too often to schedule compaction, skip it."
+                  << "compaction_type=" << compaction_task->compaction_type()
+                  << ", last_failure_time_ms=" << last_failure_ms << ", tablet_id=" << tablet->tablet_id();
+        return false;
+    }
+
+    // found a qualified tablet
+    // qualified tablet will be removed from candidates
+    need_reset_task = false;
+    return true;
+}
+
+bool CompactionScheduler::_check_precondition(const CompactionCandidate& candidate) {
+    if (!candidate.tablet) {
+        LOG(WARNING) << "null tablet";
+        return false;
+    }
+    const TabletSharedPtr& tablet = candidate.tablet;
+    if (!tablet->need_compaction(candidate.type)) {
+        // check need compaction
+        // if it is false, skip this tablet and remove it from candidate
+        LOG(INFO) << "skip tablet:" << tablet->tablet_id() << " because need_compaction is false";
+        return false;
+    }
+
+    if (tablet->tablet_state() != TABLET_RUNNING) {
+        LOG(INFO) << "skip tablet:" << tablet->tablet_id() << " because tablet state is:" << tablet->tablet_state()
+                  << ", not RUNNING";
+        return false;
+    }
+
+    // check for alter task for safety
+    // maybe this logic can be removed if tablet is the schema change dest tablet, need_compaction will never return true
+    AlterTabletTaskSharedPtr cur_alter_task = tablet->alter_task();
```

Comment:
Now, the alter is taken control by FE. AlterTask in tablet in non-sense.

Reply:
remove it.